src/libraries/System.Private.Xml/src/System/Xml/XmlResolver.FileSystemResolver.cs

```diff
@@ -0,0 +1,55 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.IO;
+using System.Threading.Tasks;
+
+namespace System.Xml
+{
+    public abstract partial class XmlResolver
+    {
+        /// <summary>
+        /// Gets an XML resolver which resolves only file system URIs.
+        /// </summary>
+        /// <value>An XML resolver which resolves only file system URIs.</value>
+        /// <remarks>
+        /// Calling <see cref="GetEntity"/> or <see cref="GetEntityAsync"/> on the
+        /// <see cref="XmlResolver"/> instance returned by this property will resolve only URIs which scheme is file.
+        /// </remarks>
+        public static XmlResolver FileSystemResolver => XmlFileSystemResolver.s_singleton;
+
+        // An XmlResolver that forbids all external entity resolution.
+        private sealed class XmlFileSystemResolver : XmlResolver
+        {
+            internal static readonly XmlFileSystemResolver s_singleton = new();
+
+            // Private constructor ensures existing only one instance of XmlFileSystemResolver
+            private XmlFileSystemResolver() { }
+
+            public override object? GetEntity(Uri absoluteUri, string? role, Type? ofObjectToReturn)
+            {
+                if ((ofObjectToReturn is null || ofObjectToReturn == typeof(Stream) || ofObjectToReturn == typeof(object))
+                    && absoluteUri.Scheme == "file")
+                {
+                    return new FileStream(absoluteUri.LocalPath, FileMode.Open, FileAccess.Read, FileShare.Read, 1);
+                }
+
+                throw new XmlException(SR.Xml_UnsupportedClass, string.Empty);
+            }
+
+            public override async Task<object> GetEntityAsync(Uri absoluteUri, string? role, Type? ofObjectToReturn)
+            {
+                if (ofObjectToReturn == null || ofObjectToReturn == typeof(Stream) || ofObjectToReturn == typeof(object))
+                {
+                    if (absoluteUri.Scheme == "file")
+                    {
+                        return await Task.Run<Stream>(() => new FileStream(absoluteUri.LocalPath, FileMode.Open, FileAccess.Read, FileShare.Read, 1, useAsync: true))
```

Comment:
I extracted this logic from XmlUrlResolver, I think we should be able to change this without issues in both places though since I don't think ctors can be async (unless there is some magic behind the back to make that work)

Reply:
@stephentoub - is there anything I'm missing here? I see you modified the XmlDownloadManager code in https://github.com/dotnet/corefx/pull/41111, but the `Task.Run<Stream>(...)` code was already there. Is there a reason to create the FileStream on a separate thread?

cc @adamsitnik

Reply:
> Is there a reason to create the FileStream on a separate thread?

Maybe someone was concerned that in these scenarios the file would be living on a remote file share that could take a non-trivial amount of time to open?